### PR TITLE
OCM-2328: Add wait attribute in cluster_rosa_classic resource

### PR DESCRIPTION
**Pull request by bardielle**

### CODE REVIEW

File: provider/clusterrosaclassic/cluster_rosa_classic_resource.go [hunk 1]
```diff
@@ -465,6 +465,10 @@ func (r *ClusterRosaClassicResource) Schema(ctx context.Context, req resource.Sc
 					PrivateHZValidator(),
 				},
 			},
+			"wait": schema.BoolAttribute{
```
Comment (nirarg) [inline provider/clusterrosaclassic/cluster_rosa_classic_resource.go:468]: Lets have more clear attribute name, something like: `wait_for_create_complete`

File: provider/clusterrosaclassic/cluster_rosa_classic_resource.go [hunk 2]
```diff
@@ -465,6 +465,10 @@ func (r *ClusterRosaClassicResource) Schema(ctx context.Context, req resource.Sc
...
+				Description: "Wait until the cluster is either in a ready state or in an error state. The waiter has a timeout of 60 minutes, with the default value set to false",
```
Comment (nirarg) [inline provider/clusterrosaclassic/cluster_rosa_classic_resource.go:469]:
```suggestion
				Description: "Wait for the cluster creation to finish. Wait for a maximum of 60 minutes, and if the creation process is not completed within this time frame, return an error. Additionally, if the creation process completes with an error state, also return an error. The default setting for this is false.",
```

File: provider/common/cluster_waiter.go [hunk 3]
```diff
@@ -22,3 +25,44 @@ func WaitTillClusterReady(ctx context.Context, collection *cmv1.ClustersClient,
...
+	defer cancel()
+	_, err := client.Poll().
+		Interval(pollingIntervalInMinutes * time.Minute).
+		Predicate(func(getClusterResponse *cmv1.ClusterGetResponse) bool {
+			object = getClusterResponse.Body()
+			tflog.Debug(ctx, "polled cluster state", map[string]interface{}{
+				"state": object.State(),
+			})
+			switch object.State() {
+			case cmv1.ClusterStateReady,
```
Comment (nirarg) [inline provider/common/cluster_waiter.go:42]: Shouldn't also check the case for ClusterSTateError in this point?

File: provider/clusterrosaclassic/cluster_rosa_classic_resource.go (hunk 1 above)
Comment (osherdp) [reply provider/clusterrosaclassic/cluster_rosa_classic_resource.go:468]: Maybe ``wait_for_ready_status``? To match the status users see for their clusters
```
$ rosa list clusters
ID                                NAME             STATE  TOPOLOGY
<id>  <name>  ready  Classic
```

Comment (nirarg) [reply provider/clusterrosaclassic/cluster_rosa_classic_resource.go:468]: I also thought about this name, but I think the purpose of this wait is to wait for any success or failure.
I don't have strong opinion about it, up to you @bardielle :-)

Comment (bardielle) [reply provider/clusterrosaclassic/cluster_rosa_classic_resource.go:468]: I'll prefer `wait_for_create_complete`

**Review by nirarg — COMMENTED** [review]
Please change the CI manifests to use the internal wait
Also please update the docs with your change

Other than that, this looks good for me

File: provider/clusterwaiter/cluster_waiter_resource.go [hunk 4]
```diff
@@ -3,14 +3,14 @@ package clusterwaiter
 import (
 	"context"
 	"fmt"
+	"github.com/terraform-redhat/terraform-provider-rhcs/provider/common"
```
Comment (tsorya) [inline provider/clusterwaiter/cluster_waiter_resource.go:6]: lets move it lower block

File: provider/common/cluster_waiter.go [hunk 5]
```diff
@@ -2,11 +2,15 @@ package common
 
 import (
 	"context"
+	"fmt"
+	"github.com/hashicorp/terraform-plugin-log/tflog"
```
Comment (tsorya) [inline provider/common/cluster_waiter.go:6]: lets move to lower block

File: provider/clusterrosaclassic/cluster_rosa_classic_resource.go [hunk 6]
```diff
@@ -1021,6 +1040,20 @@ func (r *ClusterRosaClassicResource) Update(ctx context.Context, request resourc
 		return
 	}
 
+	clusterState := "Unknown"
+	if common.HasValue(state.State) && state.State.ValueString() != "" {
```
Comment (tsorya) [inline provider/clusterrosaclassic/cluster_rosa_classic_resource.go:1044]: Wondering why we need this part. Can't we just do 
if state.State.String() != string(cmv1.ClusterStateReady) or state.State.ValueString()  != string(cmv1.ClusterStateReady) ?

File: provider/common/cluster_waiter.go [hunk 7]
```diff
@@ -22,3 +26,48 @@ func WaitTillClusterReady(ctx context.Context, collection *cmv1.ClustersClient,
...
+	if err != nil {
+		if attempts--; attempts > 0 {
+			time.Sleep(sleep)
+			return RetryClusterReadiness(attempts, 2*sleep, clusterId, ctx, timeout, clusterCollection)
+		}
+		return nil, fmt.Errorf("polling cluster state failed with error %v", err)
+	}
+
+	if object.State() == cmv1.ClusterStateError {
+		return object, fmt.Errorf("cluster creation was failed")
```
Comment (tsorya) [inline provider/common/cluster_waiter.go:69]: "cluster creation was failed" -> "cluster creation failed" ?

File: provider/machinepool/machine_pool_resource.go [hunk 8]
```diff
@@ -39,6 +34,9 @@ import (
 	"github.com/hashicorp/terraform-plugin-framework/types"
 	"github.com/hashicorp/terraform-plugin-log/tflog"
 	"github.com/terraform-redhat/terraform-provider-rhcs/provider/common"
+	"net/http"
```
Comment (tsorya) [inline provider/machinepool/machine_pool_resource.go:37]: Lets move them back to upper block

File: ci/e2e/oidc_configuration/cluster_with_managed_oidc_config/main.tf [hunk 9]
```diff
@@ -88,10 +88,5 @@ resource "rhcs_cluster_rosa_classic" "rosa_sts_cluster" {
...
   destroy_timeout = 120
-}
-
-resource "rhcs_cluster_wait" "rosa_cluster" {
-  cluster = rhcs_cluster_rosa_classic.rosa_sts_cluster.id
-  # timeout in minutes
-  timeout = 60
-}
+  wait_for_create_complete = true
+}
```
Comment (tsorya) [inline ci/e2e/oidc_configuration/cluster_with_managed_oidc_config/main.tf:92]: add new line?

File: provider/clusterrosaclassic/cluster_rosa_classic_resource.go [hunk 10]
```diff
@@ -936,6 +940,20 @@ func (r *ClusterRosaClassicResource) Create(ctx context.Context, request resourc
...
 
+	if common.HasValue(state.WaitForCreateComplete) && state.WaitForCreateComplete.ValueBool() {
+		object, err = common.RetryClusterReadiness(3, 30*time.Second, object.ID(),
+			ctx, waitTimeoutInMinutes, r.clusterCollection)
+		if err != nil {
+			response.Diagnostics.AddError(
+				"Waiting for cluster creation finished with error",
+				fmt.Sprintf("Waiting for cluster creation finished with the error %v", err),
+			)
+			if object == nil {
```
[the remark this answers is not in the document]
Comment (bardielle) [reply provider/clusterrosaclassic/cluster_rosa_classic_resource.go:951]: @nirarg had a conversation about it. 
We can populate if there is an object

File: provider/clusterrosaclassic/cluster_rosa_classic_resource.go (hunk 6 above)
Comment (bardielle) [reply provider/clusterrosaclassic/cluster_rosa_classic_resource.go:1044]: If it's an empty string I prefer the error msg will be "unknown"

**Pull request merged**